src/monmove.cpp

```diff
@@ -144,70 +166,105 @@ void monster::plan(const mfactions &factions)
             npc *me = (g->active_npc[i]);
             float rating = rate_target( *me, bresenham_slope, electronic );
-            if( rating < dist || ( !fleeing && is_fleeing( *me ) ) ) {
-                if( is_fleeing( *me ) ) {
-                    fleeing = true;
-                    set_dest(posx() * 2 - me->posx, posy() * 2 - me->posy, bresenham_slope);
-                    dist = rating;
-                } else if( attitude( me ) == MATT_ATTACK ) {
+            bool fleeing_from = is_fleeing( *me );
+            // Switch targets if closer and hostile or scarier than current target
+            if( ( rating < dist && fleeing ) ||
+                ( rating < dist && attitude( me ) == MATT_ATTACK ) ||
+                ( !fleeing && fleeing_from ) ) {
                     closest = i;
-                    selected_slope = bresenham_slope;
                     dist = rating;
-                }
+                    selected_slope = bresenham_slope;
+            }
+            fleeing = fleeing || fleeing_from;
+            if( angers_hostile_near && rating <= 5 ) {
+                anger += 5;
             }
         }
     }
 
-    if( !fleeing ) {
-        fleeing = attitude() == MATT_FLEE;
-        if( friendly == 0 && can_see() ) {
-            for( const auto &faction : factions ) {
-                if( faction.first == monfaction() ) {
-                    continue;
+    fleeing = attitude() == MATT_FLEE;
+    if( !docile && friendly == 0 ) {
+        for( const auto &fac : factions ) {
+            if( fac.first == faction->id ) {
+                continue;
+            }
+
+            for( int i : fac.second ) { // mon indices
+                monster &mon = g->zombie( i );
+                float rating = rate_target( mon, bresenham_slope, electronic );
+                if( rating < dist ) {
+                    dist = rating;
+                    closest = -3 - i;
+                    selected_slope = bresenham_slope;
+                }
+                if( angers_hostile_near && rating <= 5 ) {
+                    anger += 5;
                 }
+            }
+        }
+    }
 
-                for( int i : faction.second ) { // mon indices
-                    monster &mon = g->zombie( i );
-                    float rating = rate_target( mon, bresenham_slope, electronic );
-                    if( rating < dist ) {
-                        dist = rating;
-                        if (fleeing) {
-                            wandx = posx() * 2 - mon.posx();
-                            wandy = posy() * 2 - mon.posy();
-                            wandf = 40;
-                        } else {
-                            closest = -3 - i;
-                            selected_slope = bresenham_slope;
-                        }
-                    }
+    // Friendly monsters here
+    // Avoid for hordes of same-faction stuff or it could get expensive
+    auto myfaction = factions.find( faction->id )->second;
```

Comment:
This is inconsistent with the changes in game.cpp. If the monster is friendly, there will be an entry -1 in the factions map, but no entry with `faction->id` of the monster itself. The find will return an invalid iterator, which is dereferenced and will cause a crash.

Might be related to #10895 as it crashes at this line.